### PR TITLE
add health check for node-exporter pod 

### PR DESCRIPTION
**Pull request by YanjieGao**
- Check predefined health metrics from previous doc for gpu-exporter & node-exporter

- If readiness probe failed, from k8s page will detect the status

-  it will log the exceptions to disk.

### CODE REVIEW

File: prometheus/exporter/healthy_check.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            runTimeException.append(err)
+            logger.info(err)
+
+    try:
+        dockerInspectCMD = "docker inspect --help" 
+        dockerDockerInspect = subprocess.check_output([dockerInspectCMD], shell=True)
+    except subprocess.CalledProcessError as e:
+        err = "command '{}' return with error (code {}): {}".format(e.cmd, e.returncode, e.output)
+        runTimeException.append(err)
+        logger.info(err)
```
Comment (ydye) [inline prometheus/exporter/healthy_check.py]: Why will the error log be outputted by logger.info not logger.error?

File: prometheus/exporter/healthy_check.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+logger = logging.getLogger("node_exporter probe")  
+logger.setLevel(logging.INFO)  
+fh = RotatingFileHandler("/datastorage/prometheus/node_exporter_probe.log", maxBytes= 1024 * 1024 * 100, backupCount=5)  
+fh.setLevel(logging.INFO)  
+formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")  
+fh.setFormatter(formatter)  
+logger.addHandler(fh)  
+
+def main(argv):
```
Comment (ydye) [inline prometheus/exporter/healthy_check.py]: Remove argv

File: prometheus/exporter/healthy_check.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    if not os.path.exists("/datastorage/prometheus/job_exporter.prom"):
+        err = "/datastorage/prometheus/job_exporter.prom does not exists"
+        runTimeException.append(err)
+        logger.info(err)
+
+    if len(runTimeException) > 0:
+        raise RuntimeError("gpu-exporter readiness probe failed")
+
+if __name__ == "__main__":
+    main(sys.argv[1:])
```
Comment (ydye) [inline prometheus/exporter/healthy_check.py]: remove argv.

File: prometheus/exporter/healthy_check.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    except subprocess.CalledProcessError as e:
+        err = "command '{}' return with error (code {}): {}".format(e.cmd, e.returncode, e.output)
+        runTimeException.append(err)
+        logger.info(err)
+
+    if gpuExists:
+        try:
+            env =  os.getenv("NV_DRIVER")
+            if not env:
+                nvidiaCMD= "/bin/nvidia-smi -q -x"
```
Comment (ydye) [inline prometheus/exporter/healthy_check.py]: /bin/nvidia-smi ???? Is this the path of gpu-drivers on the container??

File: pai-management/bootstrap/prometheus/node-exporter-ds.yaml.template [hunk 5]
```diff
@@ -56,6 +62,13 @@ spec:
           name: scrape
       - image: {{ clusterinfo['dockerregistryinfo']['prefix'] }}gpu-exporter:{{ clusterinfo['dockerregistryinfo']['docker_tag'] }}
         imagePullPolicy: Always
+        readinessProbe:
+          exec:
+            command:
+            - python
+            - /usr/local/healthy_check.py
```
Comment (hwuu) [inline pai-management/bootstrap/prometheus/node-exporter-ds.yaml.template]: Could you tell me when was this file (`health_check.py`) placed into this place?

File: prometheus/exporter/healthy_check.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+logger = logging.getLogger("node_exporter probe")  
+logger.setLevel(logging.INFO)  
+fh = RotatingFileHandler("/datastorage/prometheus/node_exporter_probe.log", maxBytes= 1024 * 1024 * 100, backupCount=5)  
+fh.setLevel(logging.INFO)  
+formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")  
+fh.setFormatter(formatter)  
+logger.addHandler(fh)  
+
+def main():
+    runTimeException = []
```
Comment (hwuu) [inline prometheus/exporter/healthy_check.py]: Suggest use [PEP8 code style](https://legacy.python.org/dev/peps/pep-0008/#naming-conventions) for naming convention.

File: pai-management/bootstrap/prometheus/node-exporter-ds.yaml.template (hunk 5 above)
Comment (YanjieGao) [reply pai-management/bootstrap/prometheus/node-exporter-ds.yaml.template]: at dockerfile
https://github.com/Microsoft/pai/blob/master/pai-management/src/gpu-exporter/dockerfile

File: pai-management/bootstrap/prometheus/node-exporter-ds.yaml.template [hunk 7]
```diff
@@ -33,6 +33,12 @@ spec:
       containers:
       - image: prom/node-exporter:v0.15.2
         imagePullPolicy: Always
+        readinessProbe:
+          httpGet:
+            path: /metrics
+            port: {{ clusterinfo['prometheusinfo']['node_exporter_port'] }}
```
Comment (ydye) [inline pai-management/bootstrap/prometheus/node-exporter-ds.yaml.template]: Are there will be any server to handle this http request?

Comment (YanjieGao) [reply pai-management/bootstrap/prometheus/node-exporter-ds.yaml.template]: promethueus will fetch this address to get the metrics from every node

**Pull request merged**